### PR TITLE
Opencl opengl buffer

### PR DESCRIPTION
**Pull request by avershov**
Previously added CL/GL interop core support & sample: added ogl::Buffer support.

### CODE REVIEW

File: modules/core/src/opengl.cpp [hunk 1]
```diff
@@ -1804,4 +1804,83 @@ void convertFromGLTexture2D(const Texture2D& texture, OutputArray dst)
...
+        clAccessFlags = CL_MEM_WRITE_ONLY;
+        break;
+    }
+
+    cl_int status = 0;
+    cl_mem clBuffer = clCreateFromGLBuffer(context, clAccessFlags, buffer.bufId(), &status);
+    if (status != CL_SUCCESS)
+        CV_Error(cv::Error::OpenCLApiCallError, "OpenCL: clCreateFromGLBuffer failed");
+
+    gl::Flush();
```
Comment (alalek) [inline modules/core/src/opengl.cpp]: Why not `finish`?

File: modules/core/src/opengl.cpp [hunk 2]
```diff
@@ -1804,4 +1804,83 @@ void convertFromGLTexture2D(const Texture2D& texture, OutputArray dst)
...
+
+    status = clFinish(clQueue);
+    if (status != CL_SUCCESS)
+        CV_Error(cv::Error::OpenCLApiCallError, "OpenCL: clFinish failed");
+
+    status = clReleaseMemObject(clBuffer);
+    if (status != CL_SUCCESS)
+        CV_Error(cv::Error::OpenCLApiCallError, "OpenCL: clReleaseMemObject failed");
+#endif
+}
```
Comment (alalek) [inline modules/core/src/opengl.cpp]: Access to UMat data is invalid after this call. I believe any UMat data should be released.

File: modules/core/include/opencv2/core/opengl.hpp [hunk 3]
```diff
@@ -537,6 +537,19 @@ CV_EXPORTS void convertToGLTexture2D(InputArray src, Texture2D& texture);
  */
 CV_EXPORTS void convertFromGLTexture2D(const Texture2D& texture, OutputArray dst);
 
+/** @brief Acquire GL Buffer object to process on CL side (convert to UMat).
+@param buffer      - source Buffer object.
+@param dst         - destination UMat.
+@param accessFlags - data access flags (ACCESS_READ|ACCESS_WRITE).
+ */
+CV_EXPORTS void acquireGLBuffer(const Buffer& buffer, UMat& dst, int accessFlags);
```
Comment (alalek) [inline modules/core/include/opencv2/core/opengl.hpp]: Why is used these names: `acquireGLBuffer` and `releaseGLBuffer`?
- OpenCL uses `map/unmap`
- OpenVX uses `access/commit` (but there are allowed copy operations also)
- `acquire/release` is usually associated with resources locks, and `clEnqueueAcquireGLObjects` actually do it exactly.
But, the main function of implementation is `clCreateFromGLBuffer`, not `clEnqueueAcquireGLObjects`, so name looks very strange.

**Comment by avershov** [comment]
This is exactly a tight pair of calls, not simple "create". They are just a prolog and epilog of a single operation, which is logically acts as co-routine for user action call inside it. The main function is "create", indeed. But I failed to find suitable names for "create prolog" and "create epilog". Can you propose a better names?

**Comment by alalek** [comment]
@vpisarev Vadim, what do you think about function names: `acquireGLBuffer` and `releaseGLBuffer`?

**Comment by avershov** [comment]
How about the names "create" and "destroy"? So, it's createFromGLBuffer. But destroy...what (destroyFromGLBuffer)?

**Comment by alalek** [comment]
My suggestion for API is:

```
// namespace cv::ogl
namespace ocl {

CV_EXPORTS UMat mapGLBuffer(const Buffer& buffer, int accessFlags = ACCESS_READ | ACCESS_WRITE);

CV_EXPORTS unmapGLBuffer(UMat& u);
// in implementation use this statement without any accessFlags parameter: cl_mem clBuffer = (cl_mem)u.handle(ACCESS_READ);

} // namespace ocl
```

**Comment by alalek** [comment]
Please squash commits into one:
```
git rebase -i HEAD~9
```

**Comment by alalek** [comment]
:+1:

**Pull request merged**